File: java/sorting/QuickSortProduction.java
import java.util.Comparator;
import java.util.List;

public final class QuickSortProduction {
    private static final int INSERTION_THRESHOLD = 24;

    private QuickSortProduction() {
    }

    public static <T extends Comparable<? super T>> void quickSortInPlace(List<T> values) {
        quickSortInPlace(values, Comparator.naturalOrder());
    }

    public static <T> void quickSortInPlace(List<T> values, Comparator<? super T> comparator) {
        if (values == null || comparator == null) {
            throw new IllegalArgumentException("values and comparator must not be null");
        }

        if (values.size() < 2) {
            return;
        }

        quickRange(values, 0, values.size(), comparator);
        insertionRange(values, 0, values.size(), comparator);
    }

    private static <T> void quickRange(List<T> values, int low, int high, Comparator<? super T> comparator) {
        while (high - low > INSERTION_THRESHOLD) {
            T pivot = values.get(medianIndex(values, low, low + (high - low) / 2, high - 1, comparator));
            int left = low;
            int index = low;
            int right = high;

            while (index < right) {
                int order = comparator.compare(values.get(index), pivot);

                if (order < 0) {
                    swap(values, left++, index++);
                } else if (order > 0) {
                    swap(values, index, --right);
                } else {
                    index++;
                }
            }

            if (left - low < high - right) {
                quickRange(values, low, left, comparator);
                low = right;
            } else {
                quickRange(values, right, high, comparator);
                high = left;
            }
        }
    }

    private static <T> int medianIndex(List<T> values, int left, int mid, int right, Comparator<? super T> comparator) {
        if (comparator.compare(values.get(mid), values.get(left)) < 0) {
            int temp = left;
            left = mid;
            mid = temp;
        }

        if (comparator.compare(values.get(right), values.get(left)) < 0) {
            int temp = left;
            left = right;
            right = temp;
        }

        if (comparator.compare(values.get(right), values.get(mid)) < 0) {
            mid = right;
        }

        return mid;
    }

    private static <T> void insertionRange(List<T> values, int low, int high, Comparator<? super T> comparator) {
        for (int index = low + 1; index < high; index++) {
            T current = values.get(index);
            int position = index;

            while (position > low && comparator.compare(current, values.get(position - 1)) < 0) {
                values.set(position, values.get(position - 1));
                position--;
            }

            values.set(position, current);
        }
    }

    private static <T> void swap(List<T> values, int left, int right) {
        if (left != right) {
            T temp = values.get(left);
            values.set(left, values.get(right));
            values.set(right, temp);
        }
    }
}

File: python/sorting/quick_sort_production.py
def quick_sort_in_place(values, *, key=None, reverse=False):
    if len(values) < 2:
        return values

    key_fn = key or (lambda value: value)
    keys = [key_fn(value) for value in values]
    _quick_range(values, keys, 0, len(values), reverse)
    _insertion_range(values, keys, 0, len(values), reverse)
    return values


def _quick_range(values, keys, low, high, reverse):
    while high - low > 24:
        pivot_key = keys[_median_index(keys, low, low + (high - low) // 2, high - 1, reverse)]
        left, right = _partition(values, keys, low, high, pivot_key, reverse)

        if left - low < high - right:
            _quick_range(values, keys, low, left, reverse)
            low = right
        else:
            _quick_range(values, keys, right, high, reverse)
            high = left


def _partition(values, keys, low, high, pivot_key, reverse):
    left = low
    index = low
    right = high

    while index < right:
        if _before_key(keys[index], pivot_key, reverse):
            _swap(values, keys, left, index)
            left += 1
            index += 1
        elif _before_key(pivot_key, keys[index], reverse):
            right -= 1
            _swap(values, keys, index, right)
        else:
            index += 1

    return left, right


def _median_index(keys, left, mid, right, reverse):
    if _before_key(keys[mid], keys[left], reverse):
        left, mid = mid, left
    if _before_key(keys[right], keys[left], reverse):
        left, right = right, left
    if _before_key(keys[right], keys[mid], reverse):
        mid, right = right, mid
    return mid


def _insertion_range(values, keys, low, high, reverse):
    for index in range(low + 1, high):
        current = values[index]
        current_key = keys[index]
        position = index

        while position > low and _before_key(current_key, keys[position - 1], reverse):
            values[position] = values[position - 1]
            keys[position] = keys[position - 1]
            position -= 1

        values[position] = current
        keys[position] = current_key


def _before_key(left, right, reverse):
    return left > right if reverse else left < right


def _swap(values, keys, left, right):
    if left != right:
        values[left], values[right] = values[right], values[left]
        keys[left], keys[right] = keys[right], keys[left]
